Integrate Scalar Lazy Rotation Keccakf1600, remove register x18 and optimize stack usage

manastasova opened the pull request:
### Issues:
Resolves #CryptoAlg-1898

### Description of changes: 
Integrate Scalar Lazy Rotation assembly optimized Keccakf1600 function into AWS-LC. 
This is an integration of Armv8 optimized implementation of Keccakf1600 function based on the Lazy Rotation techniques proposed in https://eprint.iacr.org/2022/1243.pdf. The proposed design skips the rotation (rho) step during the first round  of the Keccakf1600 function, where the data is later on rotated in the following loop. The improvement is based on the free use of barrel shifter (on some platforms such as G2 and G3), where the delayed rotation is performed in the scope of the following Keccak round along with the theta (XOR instruction which provides barrel shifter) step.
The code was first contributed to https://gitlab.com/arm-research/security/pqax/-/blob/master/asm/manual/keccak_f1600/keccak_f1600_x1_scalar_asm_v5.s. There are several versions of the lazy rotation design, where the most optimal (due to reduced data dependencies) is the one integrated into AWS-LC.

This PR is composed of small changes related mainly to the stack allocation and code organization. The main changes are described in https://github.com/manastasova/aws-lc/pull/1 where the high-level SHA3 functions stack allocation design is kept as in the original aws-lc implementation (SHA3_Absorb https://github.com/aws/aws-lc/blob/c91197106c4fe78d61b45e6cdaf68e677654a4ca/crypto/fipsmodule/sha/asm/keccak1600-armv8.pl#L349 and SHA3_Squeeze https://github.com/aws/aws-lc/blob/c91197106c4fe78d61b45e6cdaf68e677654a4ca/crypto/fipsmodule/sha/asm/keccak1600-armv8.pl#L446).

The main changes are listed as:
- Reduce the stack size when possible.
- Create additional macros for easier code maintenance.
- Place the macros and definitions above the function (code) they serve.
- 

### Call-outs:
The stack allocation reserves 160B for SHA3_Absorb_lazy and 128B(+48B) for SHA3_Squeeze_lazy, where the lowest 32B are used for temporary register store inside |keccak_f1600_x1_scalar_asm_lazy_rotation| function. The rest is used for callee-saved register storage. Another 32/48B are needed to offload the arguments, which are allocated outside our macros for |SHA3_Squeeze_lazy|, thus, we denote them by +48B in the description above. The decision is made in order make the code read and maintenance easier.

### Testing:
Tests are performed based on the GTests included in AWS-LC for Keccak and SHA3/SHAKE:

- Test SHA3_224, SHA3_256, SHA3_384, and SHA3_512 as:
`./crypto/crypto_test --gtest_filter=SHA3Test.NISTTestVectors`
`./crypto/crypto_test --gtest_filter=SHA3Test.NISTTestVectors_SingleShot`
- Test SHAKE128 and SHAKE256 as:
`./crypto/crypto_test --gtest_filter=SHAKE128Test.NISTTestVectors`
`./crypto/crypto_test --gtest_filter=SHAKE256Test.NISTTestVectors`

### Performance Evaluation
There is no significant performance difference from the results reported in https://github.com/manastasova/aws-lc/pull/1.

The avg performance gain between Aws-LC(PR#898) and this PR is: 
- [16B]           8.77 - 9.09 %    Graviton 2     |      9.98 - 10.59 %     Graviton 3
- [256B]        9.25 - 9.93 %    Graviton 2     |      10.53 - 11.29 %    Graviton 3 
- [1350B]      9.49 - 10.07 %   Graviton 2     |     10.93 - 11.63 %     Graviton 3
- [8192B]      9.63 - 10.16 %   Graviton 2     |     11.06 - 11.72 %      Graviton 3
- [16384B]    9.39 - 9.89 %    Graviton 2     |     10.77 - 11.36 %     Graviton 3

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and the ISC license.

nebeid commented:
> The main changes are described in https://github.com/manastasova/aws-lc/pull/1
I understand that either PR#1 or PR#2 will be used, so I suggest making PR#2 description comprehensive, i.e. containing all the changes compared to main.

manastasova commented:
Thank you! I fixed the description and added some platform details and detailed performance results.

nebeid commented on crypto/fipsmodule/sha/asm/keccak1600_pqax-armv8_stack.pl:
```diff
@@ -0,0 +1,728 @@
+$bitstate_adr  =  "x29";
+
+$code.=<<___;
+.macro load_bitstate
```
I believe that in the original code the load and stores are interleaved with arithmetic operations, e.g. `eor`. Did you attempt that to see if it will save cycles?

manastasova replied:
The original code includes ASM implementation of _only_ the KeccakF1600 function, while in AWS-LC, the entire |SHA3_absorb| and |SHA3_squeeze| are implemented in ASM. Therefore, in AWS-LC the steps for executing |SHA3_absorb|:
  i)  load the KeccakF1600 state at the beginning of |SHA3_absorb|
  ii) absorb the input into the KeccakF1600 state
  iii) the |KeccakF1600| is executed (where the original code is interleaving with the `eor` instructions)
  iv) if the remaining input bytes to be processed are more then a block size, jump to ii)
  v) store the KeccakF1600 state
It is very similar for the |SHA3_squeeze| function.

Since larger input/output would require more then a single invocation of |KeccakF1600|, the load and store could be skipped entirely in the |KeccakF1600| (so no interleave would be possible) and could be moved to the beginning and end of |SHA3_absorb| and |SHA3_squeeze|.

nebeid replied:
For inputs longer than 1 block, I assume the code jumps from iv) to i) and load more data, right?
If that's the case, could the processing start while the data is being loaded. Maybe if attempted, it would be found that it doesn't save time and is not worth sacrificing readability.

manastasova replied:
There could be a design where the absorb (loads and XORs) is further interleaved with the XORs from the theta step inside KeccakF1600. 
My concerns here are:
i) This would require to unroll the [absorb loop](https://github.com/manastasova/aws-lc/blob/1002066a86ebf367fdf90f80f849db76ee7e76e9/crypto/fipsmodule/sha/asm/keccak1600_pqax-armv8_stack.pl#L560) to interleave with additional XORs. 
ii) The input is loaded sequentially from memory and absorbed into the bitstate in a row-by-row fashion (A[0][0] XOR input[0], A[0][1] XOR input[1]...  A[1][0] XOR input[5]...). The theta step first computes a vector of the column-wise XOR sums, therefore, accesses data in a column fashion (A[0][0] XOR A[1][0], A[0][1] XOR A[1][1], ...). Maybe the input could be accessed in a non-sequential fashion such that first A[0][0] xor in[0], A[1][0] xor in[5], etc. is being processed. This, however, would require to manually (insert add/subtract instructions) increase/decrease the input pointer.
iii) The absorb is processing (XORing) different input (block) sizes with number of bitstate lanes depending on the security levels of SHA3/SHAKE. Therefore, we would need to design different absorb for each security level/block size.
iv) Interleaving the load of bitstate with the load of input is possible for the first input block. However, the bistate load (consecutive memory) should be interrupted by loads from different memory location.

I think these changes may result in adding code complexity while not seeing speedup (maybe even worsening the latency). However, maybe I am missing something here and you are referring to a different change in the design.

nebeid replied:
Thank you for the detailed analysis, @manastasova. This was just a conceptual question, but you're more aware of the details. Since your analysis is pointing to more complexity with no obvious gains, then there is no need to go on to try it.

nebeid commented:
Are the long messages test vectors from NIST? If not, what's their origin? Maybe you can add it at the beginning of each file.

manastasova commented:
> Are the long messages test vectors from NIST? If not, what's their origin? Maybe you can add it at the beginning of each file.

Yes, they are from [NIST](https://csrc.nist.gov/Projects/cryptographic-algorithm-validation-program/Secure-Hashing#sha3vsha3vss).

manastasova commented:
Just updated the scalar ASM version with the most optimal design where only one register spill is required per KeccakF1600 round. I also added links (for all SHA3/SHAKE Test Vector files) specifying the test vector source :)

torben-hansen commented on crypto/fipsmodule/sha/asm/keccak1600_pqax-armv8_stack.pl:
```diff
@@ -0,0 +1,718 @@
+                                               }}}
+                                
+$code.=<<___;
+.asciz  "Keccak-1600 absorb and squeeze for ARMv8, CRYPTOGAMS by <appro\@openssl.org>"
```
"CRYPTOGAMS by <appro\@openssl.org>" ?

torben-hansen commented on crypto/fipsmodule/sha/asm/keccak1600_pqax-armv8_stack.pl:
```diff
@@ -0,0 +1,728 @@
+# Define |keccak_f1600_x1_scalar_asm_lazy_rotation| macros, alias, constants
+{
+$code.=<<___;
+.type   round_constants, %object
```
+1
read only data?

nebeid commented on crypto/fipsmodule/sha/asm/keccak1600_pqax-armv8_stack.pl:
```diff
@@ -0,0 +1,718 @@
+#!/usr/bin/env perl
```
I think the header of the file would need to be changed to:
- include Amazon license 
- state how it's different from the original code as in the PR description. No need to state details that can be seen directly in the code, but just high-level differences.
It's adapted from Hanno's code, but no longer the same.
